### ISSUE TITLE
ConnectionError while using dvc.api.read for local repo

### ISSUE THREAD

**Issue by karotchykau** [iterative/dvc#5081]
# Bug Report

## Description

When I use `dvc.api.read` to read a file from local repo in offline mode (without internet connection), `DVC` will try to use internet connection for some reason. Local cache is fully up-to-date.

My call:
`dvc.api.read(file_to_read, repo='../', rev='my_local_branch', remote=None, mode='r', encoding=None)`

Leads to:
```
~\Miniconda3\envs\testenv\lib\site-packages\dvc\api.py in read(path, repo, rev, remote, mode, encoding)
     92     unless a remote argument is supplied.
     93     """
---> 94     with open(
     95         path, repo=repo, rev=rev, remote=remote, mode=mode, encoding=encoding
     96     ) as fd:

~\Miniconda3\envs\testenv\lib\contextlib.py in __enter__(self)
    111         del self.args, self.kwds, self.func
    112         try:
--> 113             return next(self.gen)
    114         except StopIteration:
    115             raise RuntimeError("generator didn't yield") from None

~\Miniconda3\envs\testenv\lib\site-packages\dvc\api.py in _open(path, repo, rev, remote, mode, encoding)
     80 def _open(path, repo=None, rev=None, remote=None, mode="r", encoding=None):
     81     with _make_repo(repo, rev=rev) as _repo:
---> 82         with _repo.open_by_relpath(
     83             path, remote=remote, mode=mode, encoding=encoding
     84         ) as fd:

~\Miniconda3\envs\testenv\lib\contextlib.py in __enter__(self)
    111         del self.args, self.kwds, self.func
    112         try:
--> 113             return next(self.gen)
    114         except StopIteration:
    115             raise RuntimeError("generator didn't yield") from None

~\Miniconda3\envs\testenv\lib\site-packages\dvc\repo\__init__.py in open_by_relpath(self, path, remote, mode, encoding)
    554         try:
    555             with self.state:
--> 556                 with tree.open(
    557                     path, mode=mode, encoding=encoding, remote=remote,
    558                 ) as fobj:

~\Miniconda3\envs\testenv\lib\contextlib.py in __enter__(self)
    111         del self.args, self.kwds, self.func
    112         try:
--> 113             return next(self.gen)
    114         except StopIteration:
    115             raise RuntimeError("generator didn't yield") from None

~\Miniconda3\envs\testenv\lib\site-packages\dvc\utils\http.py in open_url(url, mode, encoding)
     14     assert mode in {"r", "rt", "rb"}
     15 
---> 16     with iter_url(url) as (response, it):
     17         bytes_stream = IterStream(it)
     18 

~\Miniconda3\envs\testenv\lib\contextlib.py in __enter__(self)
    111         del self.args, self.kwds, self.func
    112         try:
--> 113             return next(self.gen)
    114         except StopIteration:
    115             raise RuntimeError("generator didn't yield") from None

~\Miniconda3\envs\testenv\lib\site-packages\dvc\utils\http.py in iter_url(url, chunk_size)
     57             response.close()
     58 
---> 59     response = request()
     60     it = gen(response)
     61     try:

~\Miniconda3\envs\testenv\lib\site-packages\dvc\utils\http.py in request(headers)
     31     def request(headers=None):
     32         the_url = url() if callable(url) else url
---> 33         response = requests.get(the_url, stream=True, headers=headers)
     34         if response.status_code == 404:
     35             raise FileNotFoundError(f"Can't open {the_url}")

~\Miniconda3\envs\testenv\lib\site-packages\requests\api.py in get(url, params, **kwargs)
     74 
     75     kwargs.setdefault('allow_redirects', True)
---> 76     return request('get', url, params=params, **kwargs)
     77 
     78 

~\Miniconda3\envs\testenv\lib\site-packages\requests\api.py in request(method, url, **kwargs)
     59     # cases, and look like a memory leak in others.
     60     with sessions.Session() as session:
---> 61         return session.request(method=method, url=url, **kwargs)
     62 
     63 

~\Miniconda3\envs\testenv\lib\site-packages\requests\sessions.py in request(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)
    540         }
    541         send_kwargs.update(settings)
--> 542         resp = self.send(prep, **send_kwargs)
    543 
    544         return resp

~\Miniconda3\envs\testenv\lib\site-packages\requests\sessions.py in send(self, request, **kwargs)
    653 
    654         # Send the request
--> 655         r = adapter.send(request, **kwargs)
    656 
    657         # Total elapsed time of the request (approximately)

~\Miniconda3\envs\testenv\lib\site-packages\requests\adapters.py in send(self, request, stream, timeout, verify, cert, proxies)
    514                 raise SSLError(e, request=request)
    515 
--> 516             raise ConnectionError(e, request=request)
    517 
    518         except ClosedPoolError as e:

ConnectionError: HTTPSConnectionPool(host='placeholder', port=placeholder): Max retries exceeded with url: placeholder (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000002632C8B5B80>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed'))
```

### Reproduce

1. create a branch
2. `dvc add any_file`
3. `git commit the_file.dvc`
4. switch to any other branch
5. remove the file (this step might be optional)
6. turn off internet connection
7. use `dvc.api.read(path_to_this_file, repo='path_to_your_local_repo', rev='previously_created_branch', remote=None, mode='r', encoding=None)` to read the file
8. `ConnectionError` must occur

You might have to configure your remote `DVC` repo (because I have it by default).

### Expected

A file read from local repo without using internet connection.

**Output of `dvc version`:**

```console
$ dvc version
DVC version: 1.10.2 (conda)
---------------------------------
Platform: Python 3.8.5 on Windows-10-10.0.18362-SP0
Supports: azure, gdrive, gs, hdfs, http, https, s3, ssh, oss
Cache types: hardlink
Caches: local
Remotes: s3
Repo: dvc, git
```

**Comment by pared** [comment]
Context:
https://discord.com/channels/485586884165107732/563406153334128681/786887519819661332

If we don't provide `rev`, dvc gets file directly from cache, in case of `rev` present, we make the clone, and miss cache info along the way.

**Comment by karotchykau** [comment]
At the moment `DVC` handles local repos without cloning (which should resolve the current issue). However, there was the [issue#5590](https://github.com/iterative/dvc/issues/5590) that didn't allow local repos to be used. It has been fixed in the [PR#5822](https://github.com/iterative/dvc/pull/5822), therefore, this issue is resolved too.
